sunscreen/src/compiler.rs

```diff
@@ -542,17 +646,35 @@ where
         self
     }
 
-    /**
-     * Add the given FHE program for compilation.
-     */
+    /// Add the given ZKP program for compilation.
+    ///
+    /// This method _will_ accept "shared" ZKP programs, which have inputs shared from
+    /// FHE programs.
+    ///
+    /// ```
+    /// # use sunscreen::{bulletproofs::BulletproofsBackend, types::zkp::{BfvSigned, BulletproofsField, Field, FieldSpec}, fhe_program, zkp_program, zkp_var, Compiler };
+    /// # #[fhe_program(scheme = "bfv")]
+    /// # fn fhe_prog() { }
+    ///
+    /// #[zkp_program]
+    /// fn zkp_prog<F: FieldSpec>(#[shared] x: BfvSigned<F>) { }
+    ///
+    /// let app = Compiler::new()
+    ///     .fhe_program(fhe_prog)
+    ///     .zkp_backend::<BulletproofsBackend>()
+    ///     .zkp_program(zkp_prog)
+    ///     .compile()
+    ///     .unwrap();
+    /// ```
     pub fn zkp_program<F>(mut self, zkp_program_fn: F) -> Self
     where
         F: ZkpProgramFn<B> + 'static,
     {
-        self.data
-            .zkp_data_mut()
-            .zkp_program_fns
-            .push(Box::new(zkp_program_fn));
+        let zkp_data = self.data.zkp_data_mut();
+        match zkp_program_fn.into_shared() {
+            Ok(shared) => zkp_data.shared_zkp_program_fns.push(shared),
+            Err(not_shared) => zkp_data.zkp_program_fns.push(not_shared),
```

Comment:
It is a bit odd, but because the method is `into_shared`, the return type seems reasonable to me. I thought it would be weird to create an either type isomorphic to result just for this but I guess we could?

Reply:
Eh up to you.